From a9a21abfbf8fd3b128db4b19e0517ca621c20528 Mon Sep 17 00:00:00 2001
From: Rong Bao <webmaster@csmantle.top>
Date: Mon, 27 Apr 2026 15:10:34 +0800
Subject: [PATCH 09/15] Bug 2023167 - Part 5: Track WASM tail call scratch
 registers on LA and RV. r=#profiler-reviewers!

* WasmTailCallFPScratchReg: `t3` on RV and `$t3` on LA.
* WasmTailCallRAScratchReg: `ra` on RV and `$ra` on LA.

Differential Revision: https://phabricator.services.mozilla.com/D296695
---
 mozglue/baseprofiler/core/platform-linux-android.cpp |  2 ++
 mozglue/baseprofiler/core/platform.cpp               |  7 ++++---
 tools/profiler/core/platform-linux-android.cpp       |  2 ++
 tools/profiler/core/platform.cpp                     | 10 +++++++---
 4 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/mozglue/baseprofiler/core/platform-linux-android.cpp b/mozglue/baseprofiler/core/platform-linux-android.cpp
index 7299cd92a9f70f5917fbc8f712169ed25ebba5fc..e617c059e5e5fe4762fb68506ccf5e6623004255 100644
--- a/mozglue/baseprofiler/core/platform-linux-android.cpp
+++ b/mozglue/baseprofiler/core/platform-linux-android.cpp
@@ -125,11 +125,13 @@ static void PopulateRegsFromContext(Registers& aRegs, ucontext_t* aContext) {
   aRegs.mSP = reinterpret_cast<Address>(mcontext.__gregs[3]);
   aRegs.mFP = reinterpret_cast<Address>(mcontext.__gregs[22]);
   aRegs.mRA = reinterpret_cast<Address>(mcontext.__gregs[1]);
+  aRegs.mT3 = reinterpret_cast<Address>(mcontext.__gregs[15]);
 #elif defined(GP_PLAT_riscv64_linux)
   aRegs.mPC = reinterpret_cast<Address>(mcontext.__gregs[0]);  // REG_PC
   aRegs.mSP = reinterpret_cast<Address>(mcontext.__gregs[2]);
   aRegs.mFP = reinterpret_cast<Address>(mcontext.__gregs[8]);
   aRegs.mRA = reinterpret_cast<Address>(mcontext.__gregs[1]);
+  aRegs.mT3 = reinterpret_cast<Address>(mcontext.__gregs[28]);
 
 #else
 #  error "bad platform"
diff --git a/mozglue/baseprofiler/core/platform.cpp b/mozglue/baseprofiler/core/platform.cpp
index 75e03258e4af6a7be59267652afc3dac1ad80923..2c525ee7330a2990acd04541476b2b1c42e241ec 100644
--- a/mozglue/baseprofiler/core/platform.cpp
+++ b/mozglue/baseprofiler/core/platform.cpp
@@ -1141,7 +1141,7 @@ static const char* const kMainThreadName = "GeckoMain";
     defined(__aarch64__)
 #  define UNWINDING_REGS_HAVE_LR_R11
 #elif defined(GP_PLAT_loongarch64_linux) || defined(GP_PLAT_riscv64_linux)
-#  define UNWINDING_REGS_HAVE_RA
+#  define UNWINDING_REGS_HAVE_RA_T3
 #endif
 
 // The registers used for stack unwinding and a few other sampling purposes.
@@ -1170,8 +1170,9 @@ class Registers {
 #elif defined(UNWINDING_REGS_HAVE_LR_R11)
   Address mLR{nullptr};   // ARM link register, or temp for return address.
   Address mR11{nullptr};  // Temp for frame pointer.
-#elif defined(UNWINDING_REGS_HAVE_RA)
+#elif defined(UNWINDING_REGS_HAVE_RA_T3)
   Address mRA{nullptr};  // Return address register.
+  Address mT3{nullptr};  // Temp for frame pointer.
 #endif
 
 #if defined(GP_OS_linux) || defined(GP_OS_android) || defined(GP_OS_freebsd)
@@ -1492,7 +1493,7 @@ static void DoPeriodicSample(PSLockRef aLock,
 #undef UNWINDING_REGS_HAVE_R10_R12
 #undef UNWINDING_REGS_HAVE_LR_R7
 #undef UNWINDING_REGS_HAVE_LR_R11
-#undef UNWINDING_REGS_HAVE_RA
+#undef UNWINDING_REGS_HAVE_RA_T3
 
 // END sampling/unwinding code
 ////////////////////////////////////////////////////////////////////////
diff --git a/tools/profiler/core/platform-linux-android.cpp b/tools/profiler/core/platform-linux-android.cpp
index 6590d320ebd808d746a5f22a21b519da6fe7cf21..9029e0e4f94619cb8eff5b2fffc80a5498d7bc12 100644
--- a/tools/profiler/core/platform-linux-android.cpp
+++ b/tools/profiler/core/platform-linux-android.cpp
@@ -117,11 +117,13 @@ static void PopulateRegsFromContext(Registers& aRegs, ucontext_t* aContext) {
   aRegs.mSP = reinterpret_cast<Address>(mcontext.__gregs[3]);
   aRegs.mFP = reinterpret_cast<Address>(mcontext.__gregs[22]);
   aRegs.mRA = reinterpret_cast<Address>(mcontext.__gregs[1]);
+  aRegs.mT3 = reinterpret_cast<Address>(mcontext.__gregs[15]);
 #elif defined(GP_PLAT_riscv64_linux)
   aRegs.mPC = reinterpret_cast<Address>(mcontext.__gregs[0]);  // REG_PC
   aRegs.mSP = reinterpret_cast<Address>(mcontext.__gregs[2]);
   aRegs.mFP = reinterpret_cast<Address>(mcontext.__gregs[8]);
   aRegs.mRA = reinterpret_cast<Address>(mcontext.__gregs[1]);
+  aRegs.mT3 = reinterpret_cast<Address>(mcontext.__gregs[28]);
 
 #else
 #  error "bad platform"
diff --git a/tools/profiler/core/platform.cpp b/tools/profiler/core/platform.cpp
index 2da2d79c0ea12c3a79ac4d3e92a860c91c0d1505..b454b3fab0285a378ee4dc4605af770d7e399415 100644
--- a/tools/profiler/core/platform.cpp
+++ b/tools/profiler/core/platform.cpp
@@ -2086,7 +2086,7 @@ static const char* const kMainThreadName = "GeckoMain";
     defined(__aarch64__)
 #  define UNWINDING_REGS_HAVE_LR_R11
 #elif defined(GP_PLAT_loongarch64_linux) || defined(GP_PLAT_riscv64_linux)
-#  define UNWINDING_REGS_HAVE_RA
+#  define UNWINDING_REGS_HAVE_RA_T3
 #endif
 
 // The registers used for stack unwinding and a few other sampling purposes.
@@ -2115,8 +2115,9 @@ class Registers {
 #elif defined(UNWINDING_REGS_HAVE_LR_R11)
   Address mLR{nullptr};   // ARM link register, or temp for return address.
   Address mR11{nullptr};  // Temp for frame pointer.
-#elif defined(UNWINDING_REGS_HAVE_RA)
+#elif defined(UNWINDING_REGS_HAVE_RA_T3)
   Address mRA{nullptr};  // Return address register.
+  Address mT3{nullptr};  // Temp for frame pointer.
 #endif
 
 #if defined(GP_OS_linux) || defined(GP_OS_android) || defined(GP_OS_freebsd)
@@ -2264,6 +2265,9 @@ static uint32_t ExtractJsFrames(
 #elif defined(UNWINDING_REGS_HAVE_LR_R11)
       registerState.lr = aRegs.mLR;
       registerState.tempFP = aRegs.mR11;
+#elif defined(UNWINDING_REGS_HAVE_RA_T3)
+      registerState.tempRA = aRegs.mRA;
+      registerState.tempFP = aRegs.mT3;
 #endif
 
       // Non-periodic sampling passes Nothing() as the buffer write position to
@@ -3138,7 +3142,7 @@ static inline void DoPeriodicSample(
 #undef UNWINDING_REGS_HAVE_R10_R12
 #undef UNWINDING_REGS_HAVE_LR_R7
 #undef UNWINDING_REGS_HAVE_LR_R11
-#undef UNWINDING_REGS_HAVE_RA
+#undef UNWINDING_REGS_HAVE_RA_T3
 
 // END sampling/unwinding code
 ////////////////////////////////////////////////////////////////////////
-- 
2.55.0

